### ISSUE TITLE
How to create a custom keymap?

### ISSUE THREAD

**Issue by 7wells** [mjhorvath/Video-Game-Keyboard-Diagrams#11]
Hello! This is an extraordinary project of yours - thank you so much! 😄 

Is there a way to create custom keymaps, i.e. something new that does not yet exist in your database? Some games offer the players to adjust the key bindings, so they would also benefit from some possibility to freely set the keys. Please forgive me in case I habe misunderstood the concept of your tool. In any case, it is really helpful (e.g. the Blender shortcuts).

PS: Sorry that I put this here as a "issue" (which is not correct, of course), but I did not know how/where else I could ask about this.

**Comment by mjhorvath** [comment]
Hi! I'm glad you like the project!

To start a new schema, select the "Blank Starter" game, and select the "Editor" format. You then have the option to submit your schema to me or save it locally to your hard drive.

**Comment by 7wells** [comment]
Cool, I was really not aware of this possibility - thank you very much! 😄 I will try (talking with the 0 A.D. developers first) and might provide something. Thanks for your kind offer!

**Comment by 7wells** [comment]
Sorry, another question:
What does 'capnor' stand for? Just a key without modiying key, e.g. just plain '5' key without Shift, Ctrl, Alt, or anyting else?

**Comment by 7wells** [comment]
Hi Michael, I have a first draft for 0 A.D. (https://github.com/0ad/0ad). I started with camera-related things and am still learning to use your great tool. So this draft is not yet for publication but rather to let you know where I am, and maybe to allow you offering me some feedback (in case you see something I should do differently - irrespective of the 0 A.D. key bindings).

The "Submit" button is greyed out, so I currently cannot send this to you. I passed the reCAPTCHA, it timed out, and I clicked again "I am no robot", however without seeing new images. Is this normal? Am I maybe missing something else to get the "Submit" button working?

I have exported my draft as HTML and can open it (seems to look ok to me). What do I have to do to continue with it (and maybe later submit it)?

Again, thanks a lot for your nice tool and sharing your thoughts. Stay well!

Kind regards
7wells

**Comment by 7wells** [comment]
Is it possible to change from e.g. a PC/AT US 104 Key (ANSI) to a Macintosh US 109 Key (A1243) after already started creating the layout in the editor, or would I have to start from scratch?

**Comment by mjhorvath** [comment]
> Sorry, I have soem further questions:
> 
> 1. If + is to zoom in (camera), but the English keyboard has + only when also pressing Shift (with Shift, it would be =), do I have to use capnor or capshf?
> 2. How can I combine e.g. Ctrl+Shift+key?

1. Yes, list them separately.
2. You can't yet. You type those in the 'Keyboard Combinations' section. In the future I may permit it, but I have not thought of a good way to communicate this to the viewer.

**Comment by mjhorvath** [comment]
> Is it possible to change from e.g. a PC/AT US 104 Key (ANSI) to a Macintosh US 109 Key (A1243) after already started creating the layout in the editor, or would I have to start from scratch?

You cannot, I'm sorry.

**Comment by mjhorvath** [comment]
Also, note that there is an icon in the top-left of the editor that looks like two curly brackets { }.

If you click on it you get to a page where you can import/export JSON code. If you feel it is easier to edit this code in a text editor, you may use it instead.

**Comment by 7wells** [comment]
> 
> 
> > Is it possible to change from e.g. a PC/AT US 104 Key (ANSI) to a Macintosh US 109 Key (A1243) after already started creating the layout in the editor, or would I have to start from scratch?
> 
> You cannot, I'm sorry.
No problem at all, as I have learnt to use your tool and could start from scratch. :) And thanks for your other replies as well! 😄

**Comment by 7wells** [comment]
> 
> 
> Also, note that there is an icon in the top-left of the editor that looks like two curly brackets { }.
> 
> If you click on it you get to a page where you can import/export JSON code. If you feel it is easier to edit this code in a text editor, you may use it instead. This code does not (yet) care whether it is a PC or Mac keyboard.

Interesting! So a game could add a function to create a proper JSON file based on the default or user-set key bindings, which could then be imported to your tool. Hmm, I have to think more about this...

**Comment by mjhorvath** [comment]
Correct. Except importing/exporting is done manually by copying and pasting at the moment. I'm not sure yet how to do file uploads in PHP.
